test/bench/bench.cpp

```diff
@@ -59,54 +73,60 @@ struct ExecutionBenchmarkCase
+    const auto engine = create_fn();
+    engine->parse(*benchmark_case.wasm_binary);
+    const auto func_idx = engine->find_function(benchmark_case.func_name);
     if (!func_idx)
     {
         return state.SkipWithError(
             ("Function \"" + benchmark_case.func_name + "\" not found").c_str());
     }
 
-    auto instance = fizzy::instantiate(module, {});
+    engine->instantiate();
+
+    auto initial_memory = fizzy::bytes{engine->get_memory()};
 
     // TODO: Check if memory is exported or imported.
-    if (benchmark_case.memory.size() > instance.memory.size())
+    if (benchmark_case.memory.size() > initial_memory.size())
         return state.SkipWithError("Cannot init memory");
+
     std::copy(std::begin(benchmark_case.memory), std::end(benchmark_case.memory),
-        std::begin(instance.memory));
-    const auto initial_memory = instance.memory;
+        std::begin(initial_memory));
+    engine->set_memory(initial_memory);
 
     {  // Execute once and check results against expectations.
-        const auto result = fizzy::execute(instance, *func_idx, benchmark_case.func_args);
+        const auto result = engine->execute(*func_idx, benchmark_case.func_args);
         if (result.trapped)
             return state.SkipWithError("Trapped");
 
         if (benchmark_case.expected_result)
         {
-            if (result.stack.size() != 1)
-                return state.SkipWithError("Incorrect number of results");
-            else if (result.stack[0] != *benchmark_case.expected_result)
+            if (!result.value)
+                return state.SkipWithError("Missing result value");
+            else if (*result.value != *benchmark_case.expected_result)
                 return state.SkipWithError("Incorrect result");
         }
-        else if (!result.stack.empty())
+        else if (result.value)
             return state.SkipWithError("Unexpected result");
 
-        if (instance.memory.size() < benchmark_case.expected_memory.size())
+        const auto mem = engine->get_memory();
+        if (mem.size() < benchmark_case.expected_memory.size())
             return state.SkipWithError("Result memory is shorted than expected");
 
         // Compare _beginning_ segment of the memory with expected.
         // Specifying expected full memory pages is impractical.
         if (!std::equal(std::begin(benchmark_case.expected_memory),
-                std::end(benchmark_case.expected_memory), std::begin(instance.memory)))
+                std::end(benchmark_case.expected_memory), std::begin(mem)))
             return state.SkipWithError("Incorrect result memory");
     }
 
     for (auto _ : state)  // NOLINT(clang-analyzer-deadcode.DeadStores)
     {
-        instance.memory = initial_memory;  // Restore initial memory.
-        const auto result = fizzy::execute(instance, *func_idx, benchmark_case.func_args);
+        engine->set_memory(initial_memory);
```

Comment:
Perhaps a note here would be useful: "Resetting" the instance. This works, as long as we're not using globals or imports.